server/src/main/java/server/cache/WorkspaceTree.java

```diff
@@ -0,0 +1,70 @@
+package server.cache;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class WorkspaceTree<T> {
+    private Node<T> root;
```

Comment:
What's the purpose of the generics here?

Reply:
Just good practices.

Reply:
I mean in terms of usage. The generic in this case would restrict all nodes in the tree to be of type `T`. How will that tie in to querying and constructing the tree? Our tree will need to look something like this

```
Workspace(
  children: [
    BuildFile(
      children: [
        Target(),
        Target(),
      ],
    ),
     BuildFile(
      children: [
        Target(),
        Target(),
      ],
    ),
  ],
)
```

But the Workspace is constrained to nodes of type T only, so it doesn't seem like we can provide this structure?

Reply:
If you plug `Package` in for `T` then it works.

Reply:
So why not use package instead of T?

Reply:
What if we wanted a Tree of SourceFiles for some reason?

Reply:
Point ☝️  In that case we will have to generalize the queries to work for either packages or source files. @jasonxris This is going to impact your task, what do you think here?

Another thing to consider is that bazel workspaces are techincally composed of packages, source files, and targets, all in the same tree. Representing this all by one WorkspaceTree object might be easier to understand in the context of bazel

Reply:
`Package` contain a list of `BuildTarget` and `SourceFile`. So creating a `WorkspaceTree<Package>` will create a tree object that has a Bazel context.

Reply:
My main concern is that generalizing a <T> param for a workspace isn't needed. This Workspace tree is only meant for Bazel, we don't have any other uses for it. So why generalize?

If we generalize now, the api for querying a workspace will become more complicated to implement, extend, maintain, and invoke. Reason being is because we will be forced to generalize *how* we query. For example, if we want the WorkspaceTree to be queryable for packages, packages would likely have to implement some interface `IQueryable` so that when calling the workspace we could do something like

```
WorkspaceTree.findByString(<Some generalized type to query by>)

...

class Package implements IQueryable<String>, IQueryable<int>, IQueryable<BazelRule>, ... {
  ...
}
```

Point being that generalizing isn't necessary here. It will only make our lives harder. In my experience, you should **only** generalize if you have a legitimate need for it.